swarm_copy/tools/get_me_model_tool.py

```diff
@@ -0,0 +1,227 @@
+"""Module defining the Get ME Model tool."""
+
+import logging
+from typing import Any, ClassVar
+
+from pydantic import BaseModel, Field
+
+from swarm_copy.cell_types import get_celltypes_descendants
+from swarm_copy.tools.base_tool import BaseMetadata, BaseTool
+from swarm_copy.utils import get_descendants_id
+
+logger = logging.getLogger(__name__)
+
+
+class GetMEModelInput(BaseModel):
+    """Inputs of the knowledge graph API."""
+
+    brain_region_id: str = Field(
+        description="ID of the brain region of interest. To get this ID, please use the `resolve-entities-tool` first."
+    )
+    mtype_id: str | None = Field(
+        default=None,
+        description="ID of the M-type of interest. To get this ID, please use the `resolve-entities-tool` first.",
+    )
+    etype_id: str | None = Field(
+        default=None,
+        description="ID of the electrical type of the cell. Can be obtained through the 'resolve-entities-tool'.",
+    )
+
+
+class GetMEModelMetadata(BaseMetadata):
+    """Metadata class for GetMEModelTool."""
+
+    knowledge_graph_url: str
+    token: str
+    me_model_search_size: int
+    brainregion_path: str
+    celltypes_path: str
+
+
+class MEModelOutput(BaseModel):
+    """Output schema for the knowledge graph API."""
+
+    me_model_id: str
+    me_model_name: str | None
+    me_model_description: str | None
+    mtype: str | None
+    etype: str | None
+
+    brain_region_id: str
+    brain_region_label: str | None
+
+    subject_species_label: str | None
+    subject_age: str | None
+
+
+class GetMEModelTool(BaseTool):
+    """Class defining the Get ME Model logic."""
+
+    name: ClassVar[str] = "get-me-model-tool"
+    description: ClassVar[
+        str
+    ] = """Searches a neuroscience based knowledge graph to retrieve neuron morpho-electric model (ME models) names, IDs and descriptions.
+    Requires a 'brain_region_id' which is the ID of the brain region of interest as registered in the knowledge graph.
+    Optionally accepts an mtype_id and/or an etype_id.
+    The output is a list of ME models, containing:
+    - The brain region ID.
+    - The brain region name.
+    - The subject species name.
+    - The subject age.
+    - The model ID.
+    - The model name.
+    - The model description.
+    The model ID is in the form of an HTTP(S) link such as 'https://bbp.epfl.ch/data/bbp/mmb-point-neuron-framework-model/...'."""
+    input_schema: GetMEModelInput
+    metadata: GetMEModelMetadata
+
+    def run(self) -> None:
+        """Not implemented yet."""
+        pass
+
+    async def arun(self) -> list[MEModelOutput]:
```

Comment:
Do we not catch errors with something like inhouse ToolException instead of langchain's ?

Reply:
@WonderPG

Reply:
We handle the global error handling in `swarm_copy/run.py::handle_tool_call`. See the following snippet:
```python
 try:
            raw_result = await tool_instance.arun()
        except Exception as err:
            response = {
                "role": "tool",
                "tool_call_id": tool_call.id,
                "tool_name": name,
                "content": str(err),
            }
            return response, None
```